File: backend_guide/app/api/v1/recipe_steps.py
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status

try:
    from supabase import Client
except Exception:

    class Client:  # type: ignore
        def __init__(self, *args, **kwargs):
            pass


from app.api.deps import get_user_client, get_service_client, get_current_user
from app.schemas.auth import CurrentUser
from app.schemas.recipe import RecipeStep, RecipeStepCreate, RecipeStepUpdate

router = APIRouter()


@router.get("", response_model=List[RecipeStep])
def list_recipe_steps(
    limit: int = 10,
    offset: int = 0,
    current_user: CurrentUser = Depends(get_current_user),
    user_client: Client = Depends(get_user_client),
):
    """
    List recipe steps with pagination.
    Uses user-scoped client.
    """
    try:
        res = (
            user_client.from_("recipe_steps")
            .select("*")
            .range(offset, offset + limit - 1)
            .execute()
        )
        return res.data
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to list steps: {str(e)}",
        )


@router.get("/{id}", response_model=RecipeStep)
def get_recipe_step(
    id: str,
    current_user: CurrentUser = Depends(get_current_user),
    user_client: Client = Depends(get_user_client),
):
    """
    Get a single recipe step by ID.
    Uses user-scoped client.
    """
    try:
        res = user_client.from_("recipe_steps").select("*").eq("id", id).execute()
        if not res.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Step not found or access denied.",
            )
        return res.data[0]
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to fetch step: {str(e)}",
        )


@router.post("", response_model=RecipeStep, status_code=status.HTTP_201_CREATED)
def create_recipe_step(
    body: RecipeStepCreate,
    current_user: CurrentUser = Depends(get_current_user),
    service_client: Client = Depends(get_service_client),
):
    """
    Create a new recipe step.
    Requires cms_editor role. Uses service-role client.
    """
    if current_user.role != "cms_editor":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Forbidden: Requires cms_editor role.",
        )

    try:
        insert_data = body.model_dump()
        res = service_client.from_("recipe_steps").insert(insert_data).execute()
        if not res.data:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Failed to create step.",
            )
        return res.data[0]
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to create step: {str(e)}",
        )


@router.put("/{id}", response_model=RecipeStep)
def update_recipe_step(
    id: str,
    body: RecipeStepUpdate,
    current_user: CurrentUser = Depends(get_current_user),
    service_client: Client = Depends(get_service_client),
):
    """
    Update a recipe step.
    Requires cms_editor role. Uses service-role client.
    """
    if current_user.role != "cms_editor":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Forbidden: Requires cms_editor role.",
        )

    update_data = body.model_dump(exclude_unset=True)
    if not update_data:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No update fields provided.",
        )

    try:
        res = (
            service_client.from_("recipe_steps")
            .update(update_data)
            .eq("id", id)
            .execute()
        )
        if not res.data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Step not found or update failed.",
            )
        return res.data[0]
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to update step: {str(e)}",
        )


@router.delete("/{id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_recipe_step(
    id: str,
    current_user: CurrentUser = Depends(get_current_user),
    service_client: Client = Depends(get_service_client),
):
    """
    Delete a recipe step.
    Requires cms_editor role. Uses service-role client.
    """
    if current_user.role != "cms_editor":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Forbidden: Requires cms_editor role.",
        )

    try:
        service_client.from_("recipe_steps").delete().eq("id", id).execute()
        return None
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to delete step: {str(e)}",
        )

File: backend/supabase/functions/api/routes/recipe_steps.ts
import { Hono } from 'https://deno.land/x/hono@v3.11.8/mod.ts'
import { getUserClient, getServiceClient, getCurrentUser } from '../deps.ts'

const router = new Hono()

router.get('/', async (c) => {
  const userClient = getUserClient(c)
  const limit = parseInt(c.req.query('limit') || '10')
  const offset = parseInt(c.req.query('offset') || '0')

  try {
    const { data, error } = await userClient
      .from('instruction_steps')
      .select('*')
      .range(offset, offset + limit - 1)

    if (error) {
      return c.json({ detail: `Failed to list steps: ${error.message}` }, 400)
    }

    const mapped = (data || []).map(s => ({
      id: s.id,
      recipe_id: s.recipe_id,
      step_number: s.step_number,
      description: s.instruction_text,
      instruction_text: s.instruction_text,
      is_active_cooking: s.is_active_cooking,
      timer_seconds: s.timer_seconds
    }))

    return c.json(mapped)
  } catch (err) {
    return c.json({ detail: `Failed to list steps: ${(err as any).message}` }, 400)
  }
})

router.get('/:id', async (c) => {
  const id = c.req.param('id')
  const userClient = getUserClient(c)

  try {
    const { data, error } = await userClient.from('instruction_steps').select('*').eq('id', id)
    if (error || !data || data.length === 0) {
      return c.json({ detail: 'Step not found or access denied.' }, 404)
    }
    const s = data[0]
    return c.json({
      id: s.id,
      recipe_id: s.recipe_id,
      step_number: s.step_number,
      description: s.instruction_text,
      instruction_text: s.instruction_text,
      is_active_cooking: s.is_active_cooking,
      timer_seconds: s.timer_seconds
    })
  } catch (err) {
    return c.json({ detail: `Failed to fetch step: ${(err as any).message}` }, 400)
  }
})

router.post('/', async (c) => {
  const currentUser = await getCurrentUser(c)
  if (currentUser.role !== 'cms_editor') {
    return c.json({ detail: 'Forbidden: Requires cms_editor role.' }, 403)
  }

  const serviceClient = getServiceClient()
  const body = await c.req.json()

  const payload = {
    recipe_id: body.recipe_id,
    step_number: body.step_number,
    instruction_text: body.description || body.instruction_text || '',
    is_active_cooking: body.is_active_cooking || false,
    timer_seconds: body.timer_seconds || 0
  }

  try {
    const { data, error } = await serviceClient.from('instruction_steps').insert(payload).select()
    if (error || !data || data.length === 0) {
      return c.json({ detail: `Failed to create step: ${error?.message || 'Database error'}` }, 400)
    }
    const s = data[0]
    return c.json({
      id: s.id,
      recipe_id: s.recipe_id,
      step_number: s.step_number,
      description: s.instruction_text,
      instruction_text: s.instruction_text,
      is_active_cooking: s.is_active_cooking,
      timer_seconds: s.timer_seconds
    }, 201)
  } catch (err) {
    return c.json({ detail: `Failed to create step: ${(err as any).message}` }, 400)
  }
})

router.put('/:id', async (c) => {
  const id = c.req.param('id')
  const currentUser = await getCurrentUser(c)
  if (currentUser.role !== 'cms_editor') {
    return c.json({ detail: 'Forbidden: Requires cms_editor role.' }, 403)
  }

  const serviceClient = getServiceClient()
  const body = await c.req.json()
  if (!body || Object.keys(body).length === 0) {
    return c.json({ detail: 'No update fields provided.' }, 400)
  }

  const updateFields: any = {}
  if (body.recipe_id !== undefined) updateFields.recipe_id = body.recipe_id
  if (body.step_number !== undefined) updateFields.step_number = body.step_number
  if (body.description !== undefined) updateFields.instruction_text = body.description
  if (body.instruction_text !== undefined) updateFields.instruction_text = body.instruction_text
  if (body.is_active_cooking !== undefined) updateFields.is_active_cooking = body.is_active_cooking
  if (body.timer_seconds !== undefined) updateFields.timer_seconds = body.timer_seconds

  try {
    const { data, error } = await serviceClient.from('instruction_steps').update(updateFields).eq('id', id).select()
    if (error || !data || data.length === 0) {
      return c.json({ detail: 'Step not found or update failed.' }, 404)
    }
    const s = data[0]
    return c.json({
      id: s.id,
      recipe_id: s.recipe_id,
      step_number: s.step_number,
      description: s.instruction_text,
      instruction_text: s.instruction_text,
      is_active_cooking: s.is_active_cooking,
      timer_seconds: s.timer_seconds
    })
  } catch (err) {
    return c.json({ detail: `Failed to update step: ${(err as any).message}` }, 400)
  }
})

router.delete('/:id', async (c) => {
  const id = c.req.param('id')
  const currentUser = await getCurrentUser(c)
  if (currentUser.role !== 'cms_editor') {
    return c.json({ detail: 'Forbidden: Requires cms_editor role.' }, 403)
  }

  const serviceClient = getServiceClient()
  try {
    const { error } = await serviceClient.from('instruction_steps').delete().eq('id', id)
    if (error) {
      return c.json({ detail: `Failed to delete step: ${error.message}` }, 400)
    }
    c.status(204)
    return c.body(null)
  } catch (err) {
    return c.json({ detail: `Failed to delete step: ${(err as any).message}` }, 400)
  }
})

export default router
